pkg/models/sigv4.go

```diff
@@ -63,15 +64,29 @@ func (m *SigV4Middleware) signRequest(req *http.Request) (http.Header, error) {
 		return nil, err
 	}
 
-	if req.Body != nil {
-		// consume entire request body so that the signer can generate a hash from the contents
-		body, err := ioutil.ReadAll(req.Body)
-		if err != nil {
-			return nil, err
-		}
-		return signer.Sign(req, bytes.NewReader(body), "grafana", m.Config.Region, time.Now().UTC())
+	body, err := replaceBody(req)
+	if err != nil {
+		return nil, err
+	}
+
+	if strings.Contains(req.URL.RawPath, "%2C") {
+		req.URL.RawPath = rest.EscapePath(req.URL.RawPath, false)
 	}
-	return signer.Sign(req, nil, "grafana", m.Config.Region, time.Now().UTC())
+
+	// if X-Forwarded-For header is present, omit during signing step as it breaks AWS request verification
+	forwardHeader := req.Header.Get("X-Forwarded-For")
+	if forwardHeader != "" {
+		req.Header.Del("X-Forwarded-For")
+
+		header, err := signer.Sign(req, bytes.NewReader(body), awsServiceNamespace(m.Config.DatasourceType), m.Config.Region, time.Now().UTC())
+
+		// reset pre-existing X-Forwarded-For header value
+		req.Header.Set("X-Forwarded-For", forwardHeader)
```

Comment:
I think it would be much cleaner to move line 84 to 80, because you now have code executing after a potential error happens (line 81) and it requires thorough reading to see that there isn't a bug (since the standard pattern is to check and deal with the error immediately).

Reply:
Never mind, on looking it through also locally!